### PR TITLE
feat(blob): improve gas estimation and track min gas price

### PR DESCRIPTION
**Pull request by cmwaters**
## Overview

This PR incorporates the more recent gas estimation method in celestia-app. It also makes use of a new gRPC endpoint for consensus nodes that reveals the nodes min gas price. This enables the blob module to submit transactions with the correct gas price. If the node changes that price, the blob module is able to parse the error and retry using the new gas price.

Additionally, a stubbed CoreAccessor was introduced for when a core endpoint is not provided in order to return a better/more readable error, whereas previously, the CoreAccessor depended on the grpc dial to silently fail during start and any call to the state module would return a "cannot dial address" error.

It now looks like this: 

```
{
  "jsonrpc": "2.0",
  "id": 1,
  "error": {
    "code": 1,
    "message": "node is running without state access"
  }
}
```

## Checklist

- [ ] New and updated code has appropriate documentation
- [ ] New and updated code has new and/or updated testing
- [ ] Required CI checks are passing
- [ ] Visual proof for any user facing features like CLI or documentation updates
- [ ] Linked issues closed with keywords

### CODE REVIEW

File: blob/service.go
Comment (Wondertan) [inline blob/service.go:52]: I believe this should be part of the state accessor so that other services and transactions can rely on gas price tweaks introduced here. This would also remove the need for the QueryMinimuGasPrice interface method usage of which is better be implementation detail of general tx submission code path.

**Review by Wondertan — CHANGES_REQUESTED** [review]

File: blob/service.go
Comment (cmwaters) [reply blob/service.go:52]: yeah I thought about that initially. I wasn't sure if `core.Accessor` was just supposed to be some "dumb" quasi stateless access point to the consensus node or whether it was supposed to have logic within it. I guess if it's already doing nonce counting then perhaps it's okay. 

I can shift it across but we should be careful about overloading a single component with too much functionality.

**Comment by cmwaters** [comment]
Writing tests...

File: state/core_access.go
Comment (Wondertan) [inline state/core_access.go:202]: This should be synchronized with singleflight protection as SubmitPayForBlob can be called from multiple routines.
(does not the case because no nounce handling yet, but we want concurrent PFBs in the future)

File: state/core_access.go
Comment (Wondertan) [inline state/core_access.go:183]: Do you mean make the error public var?
I agree in general, but in this particular case, it's useless. It's a sanity check.

File: state/core_access.go
Comment (Wondertan) [inline state/core_access.go:229]: Let's fix https://github.com/celestiaorg/celestia-app/issues/1885, while we are on it.
This also means the comment on SubmitPayForBlob method will need to be slightly modified.

Comment (Wondertan) [inline state/core_access.go:229]: Similarly, all the transactions would benefit from retries. I am not pushing to make it in this PR, but it's not hard to get all birds with one sot either.

**Review by Wondertan — COMMENTED** [review]
How complicated it is to do gas estimations and retries for all transactions and not just for SubmitPFB?

File: state/core_access.go
Comment (cmwaters) [reply state/core_access.go:183]: More practically its good for testing

File: state/core_access.go
Comment (cmwaters) [reply state/core_access.go:229]: Yeah I can look into that after this PR

**Comment by cmwaters** [comment]
I am not able to write all the tests I would like because of the functionality that celestia-app needs to provide in their test infra. I'm going to work on this (on celestia-app side). We can still merge this if we think logically it makes sense and I can make an issue and follow up on the missing tests in a different PR

File: state/core_access.go
Comment (Wondertan) [reply state/core_access.go:185]: Then they should be private.

Comment (Wondertan) [reply state/core_access.go:185]: Let's remove the TODO until there is a clear need for such error var

File: state/core_access.go
Comment (Wondertan) [inline state/core_access.go:211]: Pls add you handle to the TODO

Comment (cmwaters) [reply state/core_access.go:211]: Can I do it in a separate PR

File: state/core_access.go
Comment (cmwaters) [reply state/core_access.go:547]: What do you mean by synced?

File: state/core_access.go
Comment (Wondertan) [inline state/core_access.go:557]: I didn't look into the impl of it, but what if we construct it only once on start?

File: state/core_access.go
Comment (Wondertan) [reply state/core_access.go:547]: synchronized

File: state/core_access.go
Comment (Wondertan) [reply state/core_access.go:211]: I am asking to make it `//TODO(@cmwaters)`. Why a separate PR?

Comment (cmwaters) [reply state/core_access.go:211]: ah yup sorry will do

File: state/core_access.go
Comment (cmwaters) [reply state/core_access.go:557]: We can if you want but I assumed this was pretty lightweight (i.e. just boilerplate code doing the encoding/decoding)

Comment (Wondertan) [reply state/core_access.go:557]: Ok then

File: state/core_access_test.go
Comment (Wondertan) [reply state/core_access_test.go:133]: Blocked on #2527

Comment (cmwaters) [reply state/core_access_test.go]: Fixed [clean up test](https://github.com/celestiaorg/celestia-node/pull/2511/commits/3724392ad1dc75485f9774a7ebaa9eca2224e787)

File: state/core_access.go
Comment (cmwaters) [reply state/core_access.go:66]: what's wrong with `mtx`? I've always called it that

**Comment by renaynay** [comment]
```
{
  "jsonrpc": "2.0",
  "id": 1,
  "error": {
    "code": 1,
    "message": "node is running without state access"
  }
}
```

File: nodebuilder/node_test.go
Comment (renaynay) [inline nodebuilder/node_test.go:46]: useless checks

File: state/core_access.go
Comment (Wondertan) [inline state/core_access.go:112]: What do?

File: nodebuilder/state/core.go
Comment (Wondertan) [inline nodebuilder/state/core.go:23]: Could you check for the exact fields that are missing? This feels wrong and fragile.

File: nodebuilder/state/module.go
Comment (Wondertan) [inline nodebuilder/state/module.go:36]: Returning nil on start and stop would be cleaner imo

File: nodebuilder/state/core_accessor_stub.go
Comment (Wondertan) [inline nodebuilder/state/core_accessor_stub.go:18]: Let's call it after module. It's first of all the stub for the module, so that users running the node without state received a nice error if access the moduler through RPC.

Comment (Wondertan) [inline nodebuilder/state/core_accessor_stub.go:18]: This should be a separate PR, tbh

File: state/metrics.go
Comment (Wondertan) [inline state/metrics.go:17]: There is a way to make FX not call this func when CoreAccess does not exist. Also this logs will be annoying, imo

File: state/core_access.go
Comment (Wondertan) [reply state/core_access.go:112]: Answered in sync

File: state/core_access.go
Comment (Wondertan) [inline state/core_access.go:72]: Link to the issue in the app

File: state/metrics.go
Comment (Wondertan) [inline state/metrics.go:17]: There is an if IIRC in other PR that does exactly this

File: nodebuilder/state/module.go
Comment (Wondertan) [inline nodebuilder/state/module.go:32]: Solved in a diff PR

File: nodebuilder/state/module.go
Comment (Wondertan) [inline nodebuilder/state/module.go:38]: Same

File: state/core_access.go
Comment (vgonkivs) [inline state/core_access.go:556]: Does not it deserve to be a separate data structure inside the coreAccesor? Imo would be much cleaner

File: state/core_access.go
Comment (vgonkivs) [inline state/core_access.go:559]: This could be exported, so we can expose it through the RPC

File: state/core_access.go
Comment (Wondertan) [reply state/core_access.go:557]: We can park this in the issue

File: state/core_access.go
Comment (Wondertan) [reply state/core_access.go:560]: We can export min gas price for sure. Care for an issue?

File: state/core_access.go
Comment (vgonkivs) [reply state/core_access.go:557]: ok

File: state/core_access.go
Comment (vgonkivs) [reply state/core_access.go:560]: yes

**Review by vgonkivs — APPROVED** [review]

**Pull request merged**
